lib/segment/src/common/rocksdb_buffered_update_wrapper.rs

```diff
@@ -57,22 +57,50 @@ impl DatabaseColumnScheduledUpdateWrapper {
         Ok(())
     }
 
+    /// Removes from `pending_updates` all results that are flushed.
+    /// If values in `pending_updates` are changed, do not remove them.
+    fn clear_flushed_updated(
+        flushed: PendingOperations,
+        pending_operations: Arc<Mutex<PendingOperations>>,
+    ) {
+        let mut pending_guard = pending_operations.lock();
+        for id in flushed.deleted {
+            pending_guard.deleted.remove(&id);
+        }
+        for (id, value) in flushed.inserted {
+            if let Some(pending_value) = pending_guard.inserted.get(&id) {
+                if *pending_value == value {
+                    pending_guard.inserted.remove(&id);
+                }
+            }
+        }
+    }
+
     pub fn flusher(&self) -> Flusher {
-        let PendingOperations { deleted, inserted } =
-            mem::take(&mut *self.pending_operations.lock());
+        let PendingOperations { deleted, inserted } = self.pending_operations.lock().clone();
+
         debug_assert!(
             inserted.keys().all(|key| !deleted.contains(key)),
             "Key to marked for insertion is also marked for deletion!"
         );
         let wrapper = self.db.clone();
+        let pending_operations_arc = self.pending_operations.clone();
+
         Box::new(move || {
-            for id in deleted {
+            for id in deleted.iter() {
```

Comment:
Do we need to cover an execution order? If a user inserts a value and deletes it after, the deletion will be performed first.

Reply:
No, this is not a problem. The operation is listed in insertion xor deletion, not in both. The above debug assertion covers this.

If the operation for a point would be both in the insert/delete list, then we would have an ordering problem, but that is not the case.

It is possible that the current flush closure inserts point 10, and then the user deletes point 10 right away. But this is also properly covered. Yes, the current flush will persist the first operation, but we will not remove the second operation from the pending list. The following flush invocation will properly persist the second operation.

Reply:
Thanks for a detailed explanation!